grouper/services/audit_log.py

```diff
@@ -0,0 +1,64 @@
+from datetime import datetime
+from enum import Enum
+from typing import TYPE_CHECKING
+
+from grouper.models.audit_log import AuditLog, AuditLogCategory
+from grouper.models.permission import Permission
+from grouper.models.user import User
+from grouper.plugin import get_plugin_proxy
+
+if TYPE_CHECKING:
+    from grouper.models.base.session import Session  # noqa: F401
+    from grouper.usecases.authorization import Authorization  # noqa: F401
+
+
+class AuditLogAction(Enum):
+    """Possible actions and descriptions for the audit log.
+
+    The logged action will be the lowercase form of the enum name, and the enum value will be used
+    as the description.
+    """
+    DISABLE_PERMISSION = "Disabled permission"
+
+
+class AuditLogService(object):
+    """Updates the audit log when changes are made."""
+
+    def __init__(self, session):
+        # type: (Session) -> None
+        self.session = session
+
+    def log_disable_permission(self, name, authorization):
+        # type: (str, Authorization) -> None
+        on_permission = Permission.get(self.session, name=name)
+        self._log(authorization, AuditLogAction.DISABLE_PERMISSION, on_permission=on_permission)
+
+    def _log(
+            self,
+            authorization,  # type: Authorization
+            action,  # type: AuditLogAction
+            on_permission,  # type: Permission
+            category=AuditLogCategory.general,  # type: AuditLogCategory
+    ):
+        # type: (...) -> None
+        """Internal method to log an action to the audit log.
+
+        All uses of AuditLog.log should be replaced with more specific entry points like the above,
+        which in turn dispatch to this private method to make the database change.  Arguments don't
+        cover all use cases yet.  This method will be expanded as further use cases are ported to
+        this service.
+        """
+        actor = User.get(self.session, name=authorization.actor)
```

Comment:
Similarly here we can get the user from UserRepository when we have that

Reply:
Right, this refactoring stops short of the storage layer.